### PR TITLE
Allow users to sign the version tag

### PR DESCRIPTION
**Pull request by eagletmt**
This patch implements #4540 .

### CODE REVIEW

File: spec/bundler/gem_helper_spec.rb [hunk 1]
```diff
@@ -253,6 +254,21 @@ def remove_push_guard(gemspec_content)
...
+        it "when user.signingkey is present" do
+          mock_build_message app_name, app_version
+          mock_confirm_message "Tagged v#{app_version} signed with ABCDEF01."
+          mock_confirm_message "Pushed git commits and tags."
+          expect(subject).to receive(:rubygem_push).with(app_gem_path.to_s)
+
+          Dir.chdir(app_path) { sys_exec("git push -u origin master", true) }
+
+          Dir.chdir(app_path) do
+            sys_exec("git config gpg.program #{File.expand_path("../../support/fake-gpg", __FILE__)}")
```
Comment (segiddins) [inline spec/bundler/gem_helper_spec.rb]: could this use `Bundler.which("echo")`?

**Comment by segiddins** [comment]
👍 other than the one tiny comment

File: spec/bundler/gem_helper_spec.rb (hunk 1 above)
Comment (eagletmt) [inline spec/bundler/gem_helper_spec.rb]: Fixed in https://github.com/bundler/bundler/pull/4743/commits/aae87129f60052bb45156474295db31aea015164

**Comment by segiddins** [comment]
@indirect r?

**Comment by rafaelfranca** [comment]
`user.signingkey` is not what usually used to know if a commit needs a sign or not. I believe `commit.gpgsign` is more accurate with the intentions to sign a tag. We could also use `tag.forceSignAnnotated` and just let git use its default behavior (that will be creating a annotated and signed that) not passing the `-a` option.

**Comment by indirect** [comment]
Git [uses this setting to sign tags already](http://stackoverflow.com/a/20628522):
> git 2.9 (June 2016) will use user.signingKey to force signing annotated tags as well as commits: commit 61c2fe0.

@eagletmt since git already signs annotated tags if `user.signingKey` is present, why is this change to Bundler needed?

@rafaelfranca `user.signingkey` is the correct config setting according to the GitHub docs: https://help.github.com/articles/telling-git-about-your-gpg-key/. Setting `commit.gpgsign` merely tells git to automatically sign each commit you create (effectively adding `-s` to all invocations of `git commit`).

**Comment by indirect** [comment]
Oops! I misunderstood the git commit, sorry.

@rafaelfranca yep, you're totally right. 😆

@eagletmt you can get this behavior by setting `git config tag.forcesignannotated true`. I think this means it does not need to be a setting in Bundler. Do you agree?

**Comment by rafaelfranca** [comment]
> I think this means it does not need to be a setting in Bundler. Do you agree?

I think we still need to remove the `-a` option that Bundler adds because that option disable the signature.

```
tag.forceSignAnnotated
           A boolean to specify whether annotated tags created should be GPG signed. If --annotate is specified
           on the command line, it takes precedence over this option.
```

**Comment by eagletmt** [comment]
I didn't know tag.forceSignAnnotated. Removing `-a` seems sufficient to me 😌 
We can remove `-a` without changing the current behavior because `-m` implies `-a`.

> If -m <msg> or -F <file> is given and -a, -s, and -u <keyid> are absent, -a is implied
> https://www.kernel.org/pub/software/scm/git/docs/git-tag.html

**Comment by eagletmt** [comment]
totally changed my patch to simply removing `-a` option.

**Comment by indirect** [comment]
👍 @homu r+

**Pull request merged**